### PR TITLE
Issue 269: Add missing 2.3 package properties

### LINKED ISSUE #269: Add missing package properties for SPDX v2.3
See https://github.com/spdx/tools-python/issues/227:
- primary package purpose
- release date
- built date
- valid until date

### PR DESCRIPTION
**Pull request by armintaenzertng**
fixes #269 

for json, yaml and xml only; tag-value and rdf will follow later

### CODE REVIEW

File: tests/test_jsonyamlxml_writer.py [hunk 1]
```diff
@@ -1,36 +1,74 @@
...
     yield temporary_file_path
-    os.remove(temporary_file_path)
+    file_with_ending = glob.glob(temporary_file_path + "*")
+    for file in file_with_ending:
+        os.remove(file)
 
 
-def test_external_package_references(temporary_file_path) -> None:
+@pytest.mark.parametrize("out_format", ['yaml', 'xml', 'json'])
+def test_external_package_references(temporary_file_path: str, out_format: str) -> None:
```
Comment (nicoweidner) [inline tests/test_jsonyamlxml_writer.py:24]: Let's remove the type annotations on tests

File: tests/test_jsonyamlxml_writer.py [hunk 2]
```diff
@@ -1,36 +1,74 @@
...
+    temporary_file_path = "temp_external_references"
     yield temporary_file_path
-    os.remove(temporary_file_path)
+    file_with_ending = glob.glob(temporary_file_path + "*")
+    for file in file_with_ending:
+        os.remove(file)
 
 
-def test_external_package_references(temporary_file_path) -> None:
+@pytest.mark.parametrize("out_format", ['yaml', 'xml', 'json'])
```
Comment (nicoweidner) [inline tests/test_jsonyamlxml_writer.py:23]: nit: `['yaml', 'xml', 'json']` can be extracted, we use it for every test

File: tests/test_jsonyamlxml_writer.py [hunk 3]
```diff
@@ -1,36 +1,74 @@
...
+
+    file_path_with_ending = temporary_file_path + "." + out_format
+    write_anything.write_file(document, file_path_with_ending, validate=False)
+
+    parsed_document: Document = parse_file(file_path_with_ending)[0]
+    written_package: Package = parsed_document.packages[0]
+
+    assert written_package.release_date == datetime(2021, 1, 1, 12, 0, 0)
+    assert written_package.built_date == datetime(2021, 1, 1, 12, 0, 0)
+    assert written_package.valid_until_date == datetime(2022, 1, 1, 12, 0, 0)
```
Comment (nicoweidner) [inline tests/test_jsonyamlxml_writer.py:74]: Better to compare against the values `package.release_date` etc

**Review by nicoweidner — COMMENTED** [review]
Looks pretty good, thanks! I had a few minor comments

File: data/SPDXJsonExample.json [hunk 4]
```diff
@@ -48,6 +48,16 @@
...
+        "externalRefs" : [ {
+      "referenceCategory" : "SECURITY",
+      "referenceLocator" : "cpe:2.3:a:pivotal_software:spring_framework:4.1.0:*:*:*:*:*:*:*",
+      "referenceType" : "cpe23Type"
+    }, {
+      "comment" : "This is the external ref for Acme",
+      "referenceCategory" : "OTHER",
+      "referenceLocator" : "acmecorp/acmenator/4.1.3-alpha",
+      "referenceType" : "http://spdx.org/spdxdocs/spdx-example-444504E0-4F89-41D3-9A0C-0305E82C3301#LocationRef-acmeforge"
+    } ],
```
[the remark this answers is not in the document]
Comment (armintaenzertng) [reply data/SPDXJsonExample.json:60]: removed that part

File: spdx/package.py [hunk 5]
```diff
@@ -159,6 +182,12 @@ def validate_files_analyzed(self, messages):
 
         return messages
 
+    def validate_primary_package_purposes(self, messages: ErrorMessages) -> ErrorMessages:
+        for purpose in self.primary_package_purposes:
+            if purpose not in PackagePurpose:
+                messages.append("Primary package purpose has a value that is not allowed!")
+        return messages
```
[the remark this answers is not in the document]
Comment (armintaenzertng) [reply spdx/package.py:189]: fixed

File: tests/test_rdf_writer.py [hunk 6]
```diff
@@ -39,6 +40,23 @@ def test_accept_provided_doc_node(temporary_file_path) -> None:
...
+    package.checksum = checksum
+
+    with open(temporary_file_path, "wb") as out:
+        writer = Writer(document, out)
+        writer.write()
+    parser = Parser(Builder(), StandardLogger())
+    with open(temporary_file_path, "r") as file:
+        parsed_document: Document = parser.parse(file)[0]
+
+    assert parsed_document is not None
```
[the remark this answers is not in the document]
Comment (armintaenzertng) [reply tests/test_rdf_writer.py:57]: removed that commit

File: tests/test_jsonyamlxml_writer.py (hunk 1 above)
Comment (armintaenzertng) [reply tests/test_jsonyamlxml_writer.py:24]: removed that commit

File: tests/test_jsonyamlxml_writer.py (hunk 2 above)
Comment (armintaenzertng) [reply tests/test_jsonyamlxml_writer.py:23]: fixed

File: tests/test_jsonyamlxml_writer.py [hunk 7]
```diff
@@ -1,36 +1,74 @@
...
+def test_primary_package_purpose(temporary_file_path: str, out_format: str):
+    document: Document = minimal_document()
+    package: Package = document.packages[0]
+    package.primary_package_purpose = PackagePurpose.FILE
+
+    file_path_with_ending = temporary_file_path + "." + out_format
+    write_anything.write_file(document, file_path_with_ending, validate=False)
+
+    parsed_document: Document = parse_file(file_path_with_ending)[0]
+    written_package: Package = parsed_document.packages[0]
```
[the remark this answers is not in the document]
Comment (armintaenzertng) [reply tests/test_jsonyamlxml_writer.py:54]: fixed

File: tests/test_jsonyamlxml_writer.py (hunk 3 above)
Comment (armintaenzertng) [reply tests/test_jsonyamlxml_writer.py:74]: fixed

File: spdx/writers/jsonyamlxml.py [hunk 8]
```diff
@@ -174,6 +174,18 @@ def create_package_info(self, package):
         if package.has_optional_field("homepage"):
             package_object["homepage"] = package.homepage.__str__()
 
+        if package.has_optional_field("primary_package_purpose"):
+            package_object["primaryPackagePurpose"] = package.primary_package_purpose.name
+
+        if package.has_optional_field("release_date"):
+            package_object["releaseDate"] = package.release_date.isoformat() + "Z"
```
[the remark this answers is not in the document]
Comment (armintaenzertng) [reply spdx/writers/jsonyamlxml.py]: rewrote the `utils.datetime_iso_format` method

File: tests/test_jsonyamlxml_writer.py (hunk 1 above)
Comment (nicoweidner) [reply tests/test_jsonyamlxml_writer.py:27]: The type hint is still there. But this is really minor....

**Review by nicoweidner — DISMISSED** [review]
Thanks for addressing my comments!

**Review by nicoweidner — APPROVED** [review]

**Pull request merged**